docs/extending/code-overview/evaluation.rst

```diff
@@ -82,19 +82,51 @@ of the Expression Tree which will match.
 
 
 
-Rewrite/Apply/Eval Process
-==========================
+Broad description of the Rewrite/Apply/Eval Process
+===================================================
 
-Expression evaluation is an iterative and recursive tranformation
-process where we apply transformation rules, apply functions
-encountered and in the evaluation of a function it may recursively go
-through a rewrite/apply/eval step for one of its subcomputations.
+The top-level evaluation performs time and evaluation process
+management for the overall evaluation process.
 
-The entire process is iterated resulting expression that comes back
-doesn't change, or we are told to stop, e.g. an error or limit was
-encountered.
+After setting up the bookkeeping for this, there is a loop performing
+a possibly recursive Rewrite/Apply/Eval step described next.  This is
+done until that step reports that no further evaluation is needed or
+possible.
 
-Here we give in broad outline a single rewrite/apply/eval step of this
+The Rewrite/Apply/Eval step has an term-rewriting phase followed by a
+function application phase; "term" here refers to a subexpression of
+the expression. [1]_
+
+The word "apply" in the title really refers to two separate kinds of
+applications: rule application and function application.
+
+The first part of the he Rewrite/Apply/Eval step, is to reorder and/or
+rewrite the expression. This transformation is directed by looking at
+properties of the Head symbols, its associated Attributes and
+rules. An attribute like ``Orderless`` can cause terms in the
```

Comment:
> Isn't `Orderless` the only attribute that makes this?

@TiagoCavalcante Looking at Mathematica Book 5 I read that `Flat` also changes the structure of an Expression.

Reply:
@rocky, I was reading again the Wikipedia entry about S-Expressions. Something that I noticed is that in the definition, the head of an S-Expression is always an atom, which is something that does not happens in WL, even before parsing:
```
Derivative[1][f][x]
```
is an expression that has `Derivative[1][f]` as Head.  Actually, in the entry about M-Expression, https://en.wikipedia.org/wiki/M-expression, it is mentioned that WL uses a version of this kind of expression instead of S-Expressions.

Reply:
Thanks for the correction. I was neither aware that this was happening nor was I aware of M-Expressions. 

It exactly these little differences though that are important to notice and mention. Otherwise it will lead to mistakes if not confusion. 

I have corrected everything in the simplest way by changing S-Expression to M-Expression. Later on I'll probably go over so that those who are familiar with S-Expressions but not M-Expression will understand. 

The difference between M-Expression and a Mathics Expression is however important. A crucial idea of a traditional S-Expression (and I suppose M-Expression), is that there is a pretty uniform data structure, a (nested) list that represents the Expression and also that evaluation everything is controlled by the head element. Also, _this is all that is going on_, i.e. there is no hidden state or side going on other than perhaps however you retrieve values for symbols.

What I would like to get across, and again this is important, is that while this may be the situation after the parse, during evaluation a Mathics Expression doesn't have to follow this as exactly. We are free to us non-uniform data structures to represent the Mathics Expression if we want. For example if I have an association, I could replace that part of the Expression with a Python Object that is exactly or like a Python Dictionary. If the expression can be accurately represented by a SymPy expression, that is acceptable too. In practice I think we'd keep the M-Expression around as a fallback, but in eval or possibly function application and even possibly rule rewrite, an alternate form could be used instead.

Reply:
> The difference between M-Expression and a Mathics Expression is however important. A crucial idea of a traditional S-Expression (and I suppose M-Expression), is that there is a pretty uniform data structure, a (nested) list that represents the Expression and also that evaluation everything is controlled by the head element. Also, _this is all that is going on_, i.e. there is no hidden state or side going on other than perhaps however you retrieve values for symbols.
> 

OK, clarified this point.

> What I would like to get across, and again this is important, is that while this may be the situation after the parse, during evaluation a Mathics Expression doesn't have to follow this as exactly. We are free to use non-uniform data structures to represent the Mathics Expression if we want. For example if I have an association, I could replace that part of the Expression with a Python Object that is exactly or like a Python Dictionary. If the expression can be accurately represented by a SymPy expression, that is acceptable too. In practice I think we'd keep the M-Expression around as a fallback, but in eval or possibly function application and even possibly rule rewrite, an alternate form could be used instead.


Regarding this other part: In principle, what we would have at each step along the evaluation are objects that share the interface either of  ``Expression`` or ``Atom``.  Probably what is missing in the hierarchy is a class attribute that marks that certain objects (Expressions, Symbols) are evaluable, but other objects (like numbers, or BoxExpressions) are not. 

Another thing that becomes apparent (at least for me) is that methods like ``evaluation`` or ``format`` or  ``boxes_to_something`` should not be member functions of ``Expression`` and ``Symbol``, but non-member functions. This would avoid all the import loops that appear in several places of the core.